### ISSUE TITLE
Invalid Session with TCPShield

### ISSUE THREAD

**Issue by Ludovix2812** [games647/FastLogin#806]
### What happened?

When i try to join the server with a premium mc account (i did /loginfast and /premium) it says invalid session. It says it when i try yo join with the custom domain (with the normal apex hosting ip and numeric ip everything works). I also use tcp shield with this custom version of fastlogin: https://github.com/games647/FastLogin/releases. Everything worked well but one day to another it started to say "invalid session".  I also tried the dev version (which shouldn't support tcpshield from what i know) and removed tcp shield, but when i try to join with the custom domain it still says "invald session".

### What did you expect?

I expect to let me join without any error

### Steps to reproduce

I don't really know, because it started one day to another. I just have a custom domain, tcp shield and this version of fastlogin: https://github.com/games647/FastLogin/releases. When i try to join it says that error

### Plugin list

https://imgur.com/fuK5jq1

### Configuration file

https://gist.github.com/Ludovix2812/8a5921e3887956cbf20640e121fe0ffa

### Server log

```shell
https://gist.github.com/Ludovix2812/64ef1494aa05a544f3492963682e4a0b
```

### Plugin version

v1.11-SNAPSHOT-4f3c8ad

### Platform

Spigot

### Relevance

- [X] I tried the latest build (build refers to development builds not necessary a release version)
- [X] I checked for existing tickets -
If there are, please vote them with a thumps reaction and not create new ones

**Comment by games647** [comment]
Where do you get the invalid session message. Inside the client? Furthermore, I need the log where you actually login, otherwise I cannot verify it.

**Comment by games647** [comment]
I increased the verbosity of log messages if it happens. Please post the server log with these new messages.

**Comment by Ludovix2812** [comment]
> I increased the verbosity of log messages if it happens. Please post the server log with these new messages.

First, i see the invalid session message in minecraft, then i updated fastlogin to the latest dev build and tried to remove all plugins other than authme and fastlogin.

https://gist.github.com/Ludovix2812/e3a86f0f0d319e9943a4a0032c5b9af1

Oh and i saw in the console that my ip changes everytime, like the latest three numbers and the numbers after the :
ex. 217.182.27.228:34319 and if i try to join again, 217.182.27.239:52797

And if i check my public ipv4 it's not the same as the ones that are writed by the console

**Comment by games647** [comment]
The IP of the player indicates that the player is behind a proxy instance. In this case it's an OVH server. That's likely the cause of this error, because Mojang checks that no proxy is involved. If TCPShield is involved, you need to install the corresponding plugin too, because it will forward the original client IP address to the server.

**Comment by Ludovix2812** [comment]
> The IP of the player indicates that the player is behind a proxy instance. In this case it's an OVH server. That's likely the cause of this error, because Mojang checks that no proxy is involved. If TCPShield is involved, you need to install the corresponding plugin too, because it will forward the original client IP address to the server.

ok so i have to install tcpshield plugin and then how can i remove the proxy from the ovh server? i also use cloudflare pages

**Comment by games647** [comment]
TCPShield sets the original IP. The OVH proxy are likely the servers from TCPShield. Cloudflare is unrelated except if you use Cloudflare Spectrum.

**Comment by Ludovix2812** [comment]
and how can i stop the OVH proxy? Is there any method to stop the ip changes?

**Comment by Ludovix2812** [comment]
> TCPShield sets the original IP. The OVH proxy are likely the servers from TCPShield. Cloudflare is unrelated except if you use Cloudflare Spectrum.

and how can i stop the OVH proxy? Is there any method to stop the ip changes?

**Comment by games647** [comment]
If a connection is established over a third-party like TCPShield, you will always see the IP address of the TCPShield server. This happens because there is no direct connection to the player, rather a connection between the Minecraft server and TCPShield where TCPShield forwards all data back and forth. The corresponding plugin will ask TCPShield for the original client IP and tell it Spigot, but the connection will still be proxied over this service. This is the feature TCPShield provides.

**Comment by Ludovix2812** [comment]
Ok but how can i fix this error?

**Comment by games647** [comment]
Install the TCPShield plugin to let FastLogin know the original IP.

**Comment by Ludovix2812** [comment]
but i installed the latest tcpshield plugin and it does that problem, if you see the screenshot that i made ingame with /plugins you can see that tcpshield is installed but it still says 'invalid session'.

**Comment by Ludovix2812** [comment]
yeah i installed the latest tcpshield version, you can also see it in the screenshot that i made with /plugins, but it still says 'invalid session'. That was the real problem

**Comment by dearminder** [comment]
I have exactly the same issue

**Comment by games647** [comment]
You can try out the newest dev build and use the recently introduced `useProxyAgnosticResolver` option.

**Comment by Ludovix2812** [comment]
well, it does let me in but the acutal autologin doesn't work, i think because the fix for tcp shield was not implemented: https://github.com/games647/FastLogin/releases/tag/tcpshield_update-1

**Comment by dearminder** [comment]
i applied this setting and what happened, first join,  register and logged in, (have my skin), left and joined and my skin dissapear, the uuids are diffrent, not sure why

**Comment by From-pErfo** [comment]
> well, it does let me in but the acutal autologin doesn't work, i think because the fix for tcp shield was not implemented: https://github.com/games647/FastLogin/releases/tag/tcpshield_update-1

I made a version with the latest code and tcpshield fixes. I'm still waiting for players who can reproduce the bug to test it
https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update

**Comment by Ludovix2812** [comment]
> 

finally, thank you so much. I installed this version of the plugin and then i tested two things.
The first one was with useProxyAgnosticResolver set to false and it still says "invalid session"
The second test was with useProxyAgnosticResolver set to true and it worked fine, i joined the server with tcpshield and the custom ip and it did the autologin. @games647 you should add this tcp shield fix to the main code, so players who use tcpshield can also use the latest dev build

**Comment by Ludovix2812** [comment]
> > well, it does let me in but the acutal autologin doesn't work, i think because the fix for tcp shield was not implemented: https://github.com/games647/FastLogin/releases/tag/tcpshield_update-1
> 
> I made a version with the latest code and tcpshield fixes. I'm still waiting for players who can reproduce the bug to test it https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update
> 
> this has a problem

finally, thank you so much. I installed this version of the plugin and then i tested two things.
The first one was with useProxyAgnosticResolver set to false and it still says "invalid session"
The second test was with useProxyAgnosticResolver set to true and it worked fine, i joined the server with tcpshield and the custom ip and it did the autologin. @games647 you should add this tcp shield fix to the main code, so players who use tcpshield can also use the latest dev build

**Comment by dearminder** [comment]
https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update
this one has bug with 192.168.0.1 ip, didnt give me session error with useProxyAgnosticResolver enabled, but loging in dont work and im getting offline UUID

**Comment by Ludovix2812** [comment]
> https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update this one has bug with 192.168.0.1 ip, didnt give me session error with useProxyAgnosticResolver enabled, but loging in dont work and im getting offline UUID

mmh, i don't really know, i just enabled useProxyAgnosticResolver and it let me join and it also did the autologin, so for me the problem is fixed

**Comment by dearminder** [comment]
> > https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update this one has bug with 192.168.0.1 ip, didnt give me session error with useProxyAgnosticResolver enabled, but loging in dont work and im getting offline UUID
> 
> mmh, i don't really know, i just enabled useProxyAgnosticResolver and it let me join and it also did the autologin, so for me the problem is fixed

did u tried it on fresh server, with cleared register plugin database (authme in my case) and cleared fastlogin database?

**Comment by Ludovix2812** [comment]
> > > https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update this one has bug with 192.168.0.1 ip, didnt give me session error with useProxyAgnosticResolver enabled, but loging in dont work and im getting offline UUID
> > 
> > 
> > mmh, i don't really know, i just enabled useProxyAgnosticResolver and it let me join and it also did the autologin, so for me the problem is fixed
> 
> did u tried it on fresh server, with cleared register plugin database (authme in my case) and cleared fastlogin database?

mmh, no, i tried with the server that was causing the "invalid session" error

**Comment by From-pErfo** [comment]
> I made a version with the latest code and tcpshield fixes. I'm still waiting for players who can reproduce the bug to test it
> 
> https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update
> 
> this has a problem

For this release... In my tests, it didn't affect Java players, but Geyser players couldn't automatically log in and didn't have the correct UUID. 

I guess it has something to do with the IP that starts with 192.

**Comment by Ludovix2812** [comment]
> > I made a version with the latest code and tcpshield fixes. I'm still waiting for players who can reproduce the bug to test it
> > https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update
> > this has a problem
> 
> For this release... In my tests, it didn't affect Java players, but Geyser players couldn't automatically log in and didn't have the correct UUID.
> 
> I guess it has something to do with the IP that starts with 192.

do you think you can fix that?

**Comment by From-pErfo** [comment]
> 你认为你能解决这个问题吗？

I don't think, I haven't touched java programming )

**Comment by Ludovix2812** [comment]
> > do you think you can fix that?
> 
> I don't think, I haven't touched java programming )

I hope @games647  will fix it

**Comment by dearminder** [comment]
> > https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update this one has bug with 192.168.0.1 ip, didnt give me session error with useProxyAgnosticResolver enabled, but loging in dont work and im getting offline UUID
> 
> It's my fault, I copied the wrong code ) now updated it: https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update But I'm still not sure if it works, still waiting for players to test it

Looks like a fix for me :) Thanks
![obraz](https://user-images.githubusercontent.com/36340787/173310263-5398fc16-dfeb-4649-bb7c-8085b2ef4274.png)

**Comment by From-pErfo** [comment]
> > https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update this one has bug with 192.168.0.1 ip, didnt give me session error with useProxyAgnosticResolver enabled, but loging in dont work and im getting offline UUID
> 
> It's my fault, I copied the wrong code ) now updated it: https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update But I'm still not sure if it works, still waiting for players to test it

I'm sure it's fixed, this version fixed my invalid session

**Comment by Ludovix2812** [comment]
> > > https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update this one has bug with 192.168.0.1 ip, didnt give me session error with useProxyAgnosticResolver enabled, but loging in dont work and im getting offline UUID
> > 
> > 
> > It's my fault, I copied the wrong code ) now updated it: https://github.com/From-pErfo/FastLogin/releases/tag/tcpshield_update But I'm still not sure if it works, still waiting for players to test it
> 
> I'm sure it's fixed, this version fixed my invalid session

does it work with 1.19?

**Comment by From-pErfo** [comment]
> does it work with 1.19?

You can try, but don't hold out hope )

**Comment by Ludovix2812** [comment]
> > does it work with 1.19?
> 
> You can try, but don't hold out hope )

yeah i tried and it doesn't work, one thing, when fastlogin for 1.19 will be released, can you make the same fix for that version?

**Comment by From-pErfo** [comment]
> yeah i tried and it doesn't work, one thing, when fastlogin for 1.19 will be released, can you make the same fix for that version?

I wish I could, but I don't have the ability and I'm also waiting for the plugin to be compatible with 1.19

**Comment by gabuch2** [comment]
Can this be upgraded with the latest changes?

**Comment by From-pErfo** [comment]
> 可以使用 FastLogin 的最新更改进行升级吗？

I don't have a 1.19 server, have you tried `autoLoginFloodgate: true`

**Comment by dearminder** [comment]
> > Can this be upgraded with FastLogin's latest changes?
> 
> I don't have a 1.19 server, have you tried `autoLoginFloodgate: true`

I Tried and it don't work, we need your fix for 1.19

**Comment by From-pErfo** [comment]
I just tried to build `games647/FastLogin` but it failed and I don't know how to fix it..
